axon-server-connector/src/main/java/org/axonframework/axonserver/connector/event/AggregateBasedAxonServerEventStorageEngine.java

```diff
@@ -176,45 +178,81 @@ private void buildMetaData(MetaData metaData, Map<String, MetaDataValue> metaDat
 
     @Override
     public MessageStream<EventMessage<?>> source(@Nonnull SourcingCondition condition) {
-        var allCriteriaStream = condition.criteria()
-                                         .flatten()
-                                         .stream()
-                                         .map(criterion -> this.eventsForCriterion(condition, criterion))
-                                         .reduce(MessageStream.empty().cast(), MessageStream::concatWith);
+        CompletableFuture<Void> endOfStreams = new CompletableFuture<>();
+        List<AggregateSource> aggregateSources = condition.criteria()
+                                                          .flatten()
+                                                          .stream()
+                                                          .map(criteria -> this.aggregateSourceForCriterion(
+                                                                  condition, criteria
+                                                          ))
+                                                          .toList();
 
-        AtomicReference<ConsistencyMarker> markerReference = new AtomicReference<>();
-        return allCriteriaStream.map(e -> {
-            ConsistencyMarker newMarker = markerReference.accumulateAndGet(
-                    e.getResource(ConsistencyMarker.RESOURCE_KEY),
-                    (m1, m2) -> m1 == null ? m2 : m1.upperBound(m2)
-            );
-            return e.withResource(ConsistencyMarker.RESOURCE_KEY, newMarker);
-        });
+        return aggregateSources.stream()
+                               .map(AggregateSource::source)
+                               .reduce(MessageStream.empty().cast(), MessageStream::concatWith)
+                               .whenComplete(() -> endOfStreams.complete(null))
+                               .concatWith(MessageStream.fromFuture(
+                                       endOfStreams.thenApply(event -> NoEventMessage.INSTANCE),
+                                       unused -> Context.with(
+                                               ConsistencyMarker.RESOURCE_KEY,
+                                               combineAggregateMarkers(aggregateSources.stream())
+                                       )
+                               ));
     }
 
-    private MessageStream<EventMessage<?>> eventsForCriterion(SourcingCondition condition, EventCriterion criterion) {
+    private AggregateSource aggregateSourceForCriterion(SourcingCondition condition, EventCriterion criterion) {
```

Comment:
All this code does seem alwullfy similar as the JPA one. I guess the JDBC and Mongo one might have to do this too.
Would it be possible to extract this to an `MultiAggregateConcatenatingStream` or similar? :thinking:

Reply:
Yep, pretty sure we can. I was being lazy for the time being to be honest :see_no_evil: 
Do you want me to reuse this behavior in this PR?

Reply:
This is fine for now, but when we introduce other variants we should extract